src/js/autocomplete.controller.js

```diff
@@ -1,542 +1,559 @@
-                // The request becomes outdated, when the new searchText has changed during the result fetching.
-                if ((searchText || '') !== (self.searchText || '')) {
-                    return;
-                }
+    /**
+     * Handles keyboard input.
+     * @param event
+     */
+    self.keydown = function (event) {
+      switch (event.keyCode) {
+        case KEY_CODE.DOWN_ARROW:
+          if (self.loading) return;
+          event.stopPropagation();
+          event.preventDefault();
+          self.index = Math.min(self.index + 1, self.itemList.length - 1);
+          updateScroll();
+          break;
+        case KEY_CODE.UP_ARROW:
+          if (self.loading) return;
+          event.stopPropagation();
+          event.preventDefault();
+          self.index = self.index<0 ? self.itemList.length - 1 : Math.max(0, self.index - 1);
+          updateScroll();
+          break;
+        case KEY_CODE.ENTER:
+          if (self.hidden || self.loading || self.index<0 || self.itemList.length<1) return;
+          if (hasSelection()) return;
+          event.stopPropagation();
+          event.preventDefault();
+          self.selectedItem = self.itemList[self.index];
+          setLoading(false);
+          break;
+      }
+    };
 
-                handleResults(matches);
-            }
+    /**
+     * Determines if the menu should be hidden.
+     * @returns {boolean}
+     */
+    function shouldHide() {
+      if (!isSearchable()) return true;    // Hide when not able to query
+      return false;            // Hide when the dropdown is not able to show.
+    }
+
+    /**
+     * Determines whether the autocomplete is able to query within the current state.
+     * @returns {boolean}
+     */
+    function isSearchable() {
+      if (self.loading) return false; // No query when query is in progress.
+      else if (hasSelection()) return false;           // No query if there is already a selection
+      else if (!hasFocus) return false;                // No query if the input does not have focus
+      return true;
+    }
+
+    /**
+     * Returns true if the autocomplete has a valid selection.
+     * @returns {boolean}
+     */
+    function hasSelection() {
+      return !!self.selectedItem;
+    }
+
+    /**
+     * Handles changes to the searchText property.
+     * @param searchText
+     * @param previousSearchText
+     */
+    function handleSearchText(searchText, previousSearchText) {
+      self.index = -1;
+      // do nothing on init
+      if (searchText === previousSearchText) return;
+      else if (self.selectedItem && self.displayProperty1) {
+        setLoading(false);
+        if (self.selectedItem[self.displayProperty1] !== searchText) {
+          self.selectedItem = null;
+          self.hidden = shouldHide();
+          if (self.itemList && self.itemListCopy) {
+            self.itemList = angular.copy(self.itemListCopy);
```

Comment:
why are you storing the contents back to original list?

Reply:
explained here https://github.com/ShoppinPal/Materializecss-Autocomplete/pull/22#discussion_r137966183